### PR TITLE
Add Array UDF ngrams

### PR DESCRIPTION
**Pull request by chanllen**
Add ngrams udf function and unit tests, description
Create the  pr based on the discussion in  #11449

Example:
`ngrams(ARRAY['foo', 'bar', 'baz'], 2) ==> ARRAY[ARRAY['foo', 'bar'], ARRAY['bar', 'baz']]`

### CODE REVIEW

**Comment by chanllen** [comment]
@tdcmeehan Could you review this? Thanks!!

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayNgramsFunction.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+@ScalarFunction("ngrams")
+@Description("Return N-grams for the input")
+public final class ArrayNgramsFunction
+{
+    private ArrayNgramsFunction() {}
+
+    @TypeParameter("T")
+    @SqlType("array(array(T))")
+    public static Block ngrams(@TypeParameter("T") Type type, @SqlType("array(T)") Block array, @SqlType(StandardTypes.BIGINT) long n)
```
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayNgramsFunction.java]: Consider putting the type parameter and array type construction in the constructor, like this:

```java
    private final Type type;
    private final ArrayType arrayType;

    @TypeParameter("T")
    public ArrayNgramsFunction(@TypeParameter("T") Type type) {
        this.type = type;
        this.arrayType = new ArrayType(type);
    }
```

And then this method becomes an instance method (remove static modifier).

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/TestArrayNgramsFunction.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+                ImmutableList.of(2.1, 3.1)));
+        assertFunction("ngrams(ARRAY[true, false, true], 2)", new ArrayType(new ArrayType(BOOLEAN)), ImmutableList.of(
+                ImmutableList.of(true, false),
+                ImmutableList.of(false, true)));
+
+        assertFunction("ngrams(ARRAY['\u4FE1\u5FF5\u7231', '\u5E0C\u671B', '\u671B'], 2)", new ArrayType(new ArrayType(createVarcharType(3))), ImmutableList.of(
+                ImmutableList.of("\u4FE1\u5FF5\u7231", "\u5E0C\u671B"),
+                ImmutableList.of("\u5E0C\u671B", "\u671B")));
+
+        assertFunction("ngrams(ARRAY[''], 2)", new ArrayType(new ArrayType(createVarcharType(0))), ImmutableList.of(
```
Comment (tdcmeehan) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/TestArrayNgramsFunction.java]: The tests look great, can we also add a test for the empty array?

File: presto-docs/src/main/sphinx/functions/array.rst [hunk 3]
```diff
@@ -129,7 +129,11 @@ Array Functions
 
     Flattens an ``array(array(T))`` to an ``array(T)`` by concatenating the contained arrays.
 
-.. function:: reduce(array<T>, initialState S, inputFunction<S,T,S>, outputFunction<S,R>) -> R
+.. function:: ngrams(array<T>, n) -> array<array<T>>
+
+    Returns ``n``-grams for the ``array``.
```
Comment (dain) [inline presto-docs/src/main/sphinx/functions/array.rst]: Please add some examples.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayNgramsFunction.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        n = n > array.getPositionCount() ? array.getPositionCount() : n;
+        ArrayType arrayType = new ArrayType(type);
+        BlockBuilder parts = arrayType.createBlockBuilder(null, array.getPositionCount() - (int) n + 1);
+
+        for (int i = 0; i <= array.getPositionCount() - n; i++) {
+            BlockBuilder blockBuilder = type.createBlockBuilder(null, (int) n);
+            for (int j = i; j < i + n; j++) {
+                type.appendTo(array, j, blockBuilder);
+            }
+            arrayType.writeObject(parts, blockBuilder.build());
```
Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayNgramsFunction.java]: I'd use a dictionary block instead of copying all of the data.

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/TestArrayNgramsFunction.java [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+                ImmutableList.of("foo", "baz"),
+                ImmutableList.of("baz", "foo")));
+        assertFunction("ngrams(ARRAY['bar', 'foo', 'baz', 'foo'], 3)", new ArrayType(new ArrayType(createVarcharType(3))), ImmutableList.of(
+                ImmutableList.of("bar", "foo", "baz"),
+                ImmutableList.of("foo", "baz", "foo")));
+        assertFunction("ngrams(ARRAY['bar', 'foo', 'baz', 'foo'], 4)", new ArrayType(new ArrayType(createVarcharType(3))), ImmutableList.of(
+                ImmutableList.of("bar", "foo", "baz", "foo")));
+        assertFunction("ngrams(ARRAY['bar', 'foo', 'baz', 'foo'], 5)", new ArrayType(new ArrayType(createVarcharType(3))), ImmutableList.of(
+                ImmutableList.of("bar", "foo", "baz", "foo")));
+        assertFunction("ngrams(ARRAY['bar', 'foo', 'baz', 'foo'], 6)", new ArrayType(new ArrayType(createVarcharType(3))), ImmutableList.of(
```
Comment (dain) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/TestArrayNgramsFunction.java]: Add a test with a very large `N`

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/TestArrayNgramsFunction.java [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+                asList(null, null),
+                asList(null, null)));
+
+        assertFunction("ngrams(ARRAY [NULL, 3, NULL], 2)", new ArrayType(new ArrayType(INTEGER)), ImmutableList.of(
+                asList(null, 3),
+                asList(3, null)));
+    }
+
+    @Test
+    public void testTypeCombinations()
```
Comment (dain) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/TestArrayNgramsFunction.java]: Add a test with an array of arrays

**Comment by dain** [comment]
@martint can you review the syntax here?

**Comment by chanllen** [comment]
Thank you for reviewing my PR!!!

**Comment by tdcmeehan** [comment]
LGTM, but let's get feedback @dain and @martint on the syntax.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayNgramsFunction.java [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+
+        // n should not be larger than the array length
+        n = min(array.getPositionCount(), n);
+        BlockBuilder parts = arrayType.createBlockBuilder(null, array.getPositionCount() - (int) n + 1);
+        int[] ids = new int[array.getPositionCount()];
+        for (int position = 0; position < ids.length; position++) {
+            ids[position] = position;
+        }
+        for (int i = 0; i <= array.getPositionCount() - n; i++) {
+            arrayType.writeObject(parts, array.getPositions(ids, i, (int) n));
```
Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayNgramsFunction.java]: This should be done with a dictionary block to avoid copying.  Create a dictionary block using the IDs array, which will form the elements of the array, and then use `ArrayBlock.fromElementsBlock` to turn the elements (dictionary block) in to an array block.

Comment (chanllen) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayNgramsFunction.java]: Make sense!. I try to use the code below but it ends up in an infinite loop calling ` private ArrayBlock(int arrayOffset, int positionCount, boolean[] valueIsNull, int[] offsets, Block values)`
I should miss something obvious....
Is this the right way? 
```
ngrams(ARRAY['bar', 'foo', 'baz', 'foo'], 2)
```
```
        // n should not be larger than the array length
        n = min(array.getPositionCount(), n);
        int numOfRecords = array.getPositionCount() - (int) n + 1;
        int[] ids = new int[numOfRecords * (int) n];
        for (int i = 0; i < numOfRecords; i += 1) {
            for (int j = i; j < n + i; j++) {
                ids[i + j] = j;
            }
        }
       
        System.out.println(Arrays.toString(ids));  //ids = [0, 1, 1, 2, 2, 3]
        int[] offset = new int[numOfRecords + 1];
        for (int i = 0; i < offset.length; i++) {
            offset[i] = i * (int) n;
        }
       System.out.println(Arrays.toString(offset)); // offset = [0, 2, 4]
        
        return ArrayBlock.fromElementBlock(numOfRecords, new boolean[numOfRecords], offset, new DictionaryBlock(array, ids));
```

Comment (dain) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayNgramsFunction.java]: It looks like you are using an old version of Presto.  The signature for that method is now (https://github.com/prestodb/presto/blob/master/presto-spi/src/main/java/com/facebook/presto/spi/block/ArrayBlock.java#L45):
```java
public static Block fromElementBlock(int positionCount, Optional<boolean[]> valueIsNull, int[] arrayOffset, Block values)
```
so the second argument should be `Optional.empty()`.

As for the infinite loop, use a debugger to figure out where it is looping forever.  If it isn't obvious, post the relevant part of the stack.

Comment (chanllen) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayNgramsFunction.java]: Will look deeper into this but want to post here in case there is any beforehand knowledge
Can we use fromElementBlock upon the DictionaryBlock? I have the following observations.


Block.getPositionCount returns the DictionaryBlock while Block (VariableWidthBlock).copyPositionCount returns VariableWidthBlock

When I change the return statement to
`ArrayBlock.fromElementBlock(numOfRecords, Optional.empty(), offset, array.copyPositions(ids, 0, numOfRecords * (int) n));`
 it works perfectly. but 
`ArrayBlock.fromElementBlock(numOfRecords, Optional.empty(), offset, array.getPositions(ids, 0, numOfRecords * (int) n));` Fails with the following exception while keeping calling the ArrayBlock constructor and ArrayBlock.createArrayBlockInternal

```
com.facebook.presto.spi.PrestoException: The optimizer exhausted the time limit of 180000 ms

	at com.facebook.presto.sql.planner.iterative.IterativeOptimizer$Context.checkTimeoutNotExhausted(IterativeOptimizer.java:277)
	at com.facebook.presto.sql.planner.iterative.IterativeOptimizer.exploreNode(IterativeOptimizer.java:127)
	at com.facebook.presto.sql.planner.iterative.IterativeOptimizer.exploreGroup(IterativeOptimizer.java:103)
	at com.facebook.presto.sql.planner.iterative.IterativeOptimizer.exploreChildren(IterativeOptimizer.java:185)
	at com.facebook.presto.sql.planner.iterative.IterativeOptimizer.exploreGroup(IterativeOptimizer.java:105)
	at com.facebook.presto.sql.planner.iterative.IterativeOptimizer.exploreChildren(IterativeOptimizer.java:185)
	at com.facebook.presto.sql.planner.iterative.IterativeOptimizer.exploreGroup(IterativeOptimizer.java:105)
	at com.facebook.presto.sql.planner.iterative.IterativeOptimizer.exploreChildren(IterativeOptimizer.java:185)
	at com.facebook.presto.sql.planner.iterative.IterativeOptimizer.exploreGroup(IterativeOptimizer.java:105)
	at com.facebook.presto.sql.planner.iterative.IterativeOptimizer.exploreChildren(IterativeOptimizer.java:185)
	at com.facebook.presto.sql.planner.iterative.IterativeOptimizer.exploreGroup(IterativeOptimizer.java:105)
	at com.facebook.presto.sql.planner.iterative.IterativeOptimizer.optimize(IterativeOptimizer.java:94)
	at com.facebook.presto.sql.planner.LogicalPlanner.plan(LogicalPlanner.java:158)
	at com.facebook.presto.testing.LocalQueryRunner.createPlan(LocalQueryRunner.java:838)
	at com.facebook.presto.testing.LocalQueryRunner.createPlan(LocalQueryRunner.java:783)
	at com.facebook.presto.testing.LocalQueryRunner.createPlan(LocalQueryRunner.java:774)
	at com.facebook.presto.testing.LocalQueryRunner.createPlan(LocalQueryRunner.java:769)
	at com.facebook.presto.testing.LocalQueryRunner.createDrivers(LocalQueryRunner.java:665)
	at com.facebook.presto.testing.LocalQueryRunner.executeInternal(LocalQueryRunner.java:620)
	at com.facebook.presto.testing.LocalQueryRunner.lambda$execute$4(LocalQueryRunner.java:590)
	at com.facebook.presto.transaction.TransactionBuilder.execute(TransactionBuilder.java:150)
	at com.facebook.presto.testing.LocalQueryRunner.inTransaction(LocalQueryRunner.java:602)
	at com.facebook.presto.testing.LocalQueryRunner.execute(LocalQueryRunner.java:590)
	at com.facebook.presto.testing.LocalQueryRunner.execute(LocalQueryRunner.java:584)
	at com.facebook.presto.operator.scalar.FunctionAssertions.executeProjectionWithAll(FunctionAssertions.java:582)
	at com.facebook.presto.operator.scalar.FunctionAssertions.selectUniqueValue(FunctionAssertions.java:304)
	at com.facebook.presto.operator.scalar.FunctionAssertions.selectSingleValue(FunctionAssertions.java:299)
	at com.facebook.presto.operator.scalar.FunctionAssertions.assertFunction(FunctionAssertions.java:262)
	at com.facebook.presto.operator.scalar.AbstractTestFunctions.assertFunction(AbstractTestFunctions.java:94)
	at com.facebook.presto.operator.scalar.TestArrayNgramsFunction.testTmp(TestArrayNgramsFunction.java:32)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:645)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:851)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1177)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112)
	at org.testng.TestRunner.privateRun(TestRunner.java:756)
	at org.testng.TestRunner.run(TestRunner.java:610)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:387)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:382)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:340)
	at org.testng.SuiteRunner.run(SuiteRunner.java:289)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1293)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1218)
	at org.testng.TestNG.runSuites(TestNG.java:1133)
	at org.testng.TestNG.run(TestNG.java:1104)
	at org.testng.IDEARemoteTestNG.run(IDEARemoteTestNG.java:72)
	at org.testng.RemoteTestNGStarter.main(RemoteTestNGStarter.java:123)
```

Comment (dain) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayNgramsFunction.java]: Yes it should work.  These abstractions should work with any block, otherwise it is a bug.  I'd ask on Slack for help with determining what is happening in the optimizer.

Comment (tdcmeehan) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayNgramsFunction.java]: @dain I did a little digging into why @chanllen's changes aren't working, let me paraphrase what the problem is.  The output of this function when using the technique above is an ArrayBlock whose inner `values` are a `DictionaryBlock`.  When we write the `ArrayBlock` out, that will `getRegion` on the inner `DictionaryBlock`, we'll write out that view of the inner `DictionaryBlock` and then we'll write out the offsets to the inner dictionary block.  When a dictionary block is written, we write the metadata of the dictionary out, in addition to its value.  One of the metadata values is the sequence ID, which in this case is just an auto-incrementing atomic long.  Because this whole thing is a literal expression, in the optimizer, we check to see the output that we wrote is different from the output we previously wrote.  The optimizer will only finish when the outputs don't differ (I presume this because if they don't differ it presumes it could optimize the output further).  Because the sequence ID always prevents this check from succeeding, because it is written out with the rest of the dictionary, it gets stuck in an infinite loop trying to optimize it forever, and is eventually killed.

A really simple way to fix this problem is to change the `getRegion` method in `DictionaryBlock` to return the parent's `dictionarySourceId` rather than generate a new one.  Then everything works, because we get deterministic output (no auto-incrementing part which changes on every `getRegion` call).  I think this should be safe to do because `getRegion` is considered a view on top of an existing dictionary, so it would share the same ID as the parent, whereas `copyRegion` is creating a copy.  It's the same reason existing method `getPositions` returns the parent's dictionary ID whereas `copyPositions` creates a new one.

For demonstration, I rebased @chanllen's branch and added this commit which passes the tests: https://github.com/tdcmeehan/presto/commit/9d3e20735f575f395f9473f9b92973131cfb829f#diff-848d6c442b31fbba6d6fcd07885a5680

Comment (chanllen) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayNgramsFunction.java]: @tdcmeehan Thank you very much!! Do you want to create a diff to fix the DictionaryBlock? I think it would be better to fix it in another standalone diff if it makes sense to you. Or I can fix it in this diff.

Comment (dain) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayNgramsFunction.java]: @tdcmeehan that sounds correct to me.  

@martint and @findepi, you might be interested in Tim's analysis above.

Comment (tdcmeehan) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayNgramsFunction.java]: @chanllen, since it's a one line change, feel free to just add it as a separate commit to this PR.  I think you'll also need to rebase against the latest master since your branch is badly out of date, let me know if you need help doing that.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayNgramsFunction.java [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+    @TypeParameter("T")
+    @SqlType("array(array(T))")
+    public static Block ngrams(@SqlType("array(T)") Block array, @SqlType(StandardTypes.BIGINT) long n)
+    {
+        checkCondition(n > 0, INVALID_FUNCTION_ARGUMENT, "N must be positive");
+        checkCondition(n <= Integer.MAX_VALUE, INVALID_FUNCTION_ARGUMENT, "N is too large");
+
+        // n should not be larger than the array length
+        int numOfRecords = toIntExact(min(array.getPositionCount(), n));
```
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayNgramsFunction.java]: I think I used the wrong name here for this variable.  Perhaps this should be `recordsPerElement`?

File: presto-spi/src/main/java/com/facebook/presto/spi/block/DictionaryBlock.java [hunk 9]
```diff
@@ -294,7 +294,7 @@ public Block copyPositions(int[] positions, int offset, int length)
     public Block getRegion(int positionOffset, int length)
     {
         checkValidRegion(positionCount, positionOffset, length);
-        return new DictionaryBlock(idsOffset + positionOffset, length, dictionary, ids, false, randomDictionaryId());
+        return new DictionaryBlock(idsOffset + positionOffset, length, getDictionary(), ids, false, getDictionarySourceId());
```
Comment (tdcmeehan) [inline presto-spi/src/main/java/com/facebook/presto/spi/block/DictionaryBlock.java]: Let's split this into its own commit, and squash the rest.

**Comment by dain** [comment]
Superseded by #11727
